# Week 2: Classification Model Training and Evaluation

## Step 1: Loading Cleaned Data and Splitting Features ($X$) and Target ($y$)
In this step, we load the cleaned Titanic dataset. We then separate our dataset into:
* **Features ($X$):** The input variables we will use to predict survival (Passenger Class, Gender, Age, and Fare).
* **Target ($y$):** The binary outcome variable we want to predict (`Survived`).

In [2]:
import pandas as pd

# 1. Load your cleaned dataset from Week 1
# (Make sure 'train.csv' is in your folder, or use the exact name of your cleaned CSV)
df = pd.read_csv('D:\Vortex Internship\Titanic-Dataset.csv') 

# Clean up basic missing values as we did in Week 1 so the model doesn't crash
df['Age'] = df['Age'].fillna(df['Age'].median())
df['Embarked'] = df['Embarked'].fillna(df['Embarked'].mode()[0])

# Preprocess: Convert 'Sex' to numerical values (0 for male, 1 for female)
# Machine Learning models require numbers, not text!
df['Sex'] = df['Sex'].map({'male': 0, 'female': 1})

# 2. Identify your Target (y) and Features (X)
y = df['Survived']                        # The target we want to predict
X = df[['Pclass', 'Sex', 'Age', 'Fare']]   # The inputs/features we will use

# 3. Print shapes to verify
print("Features shape (X):", X.shape)
print("Target shape (y):", y.shape)
print("\nFirst 5 rows of features (X):")
display(X.head())

Features shape (X): (891, 4)
Target shape (y): (891,)

First 5 rows of features (X):


<>:5: SyntaxWarning: invalid escape sequence '\V'
<>:5: SyntaxWarning: invalid escape sequence '\V'
C:\Users\PMLS\AppData\Local\Temp\ipykernel_6828\3338095773.py:5: SyntaxWarning: invalid escape sequence '\V'
  df = pd.read_csv('D:\Vortex Internship\Titanic-Dataset.csv')


,Pclass,Sex,Age,Fare
0,3,0,22.0,7.2500
1,1,1,38.0,71.2833
2,3,1,26.0,7.9250
3,1,1,35.0,53.1000
4,3,0,35.0,8.0500


## Step 2: Preprocessing, Train-Test Split, and Model Initialization
In this step, we:
1. Convert the categorical `Embarked` feature into numerical dummy variables using `pd.get_dummies`.
2. Split our features ($X$) and target ($y$) into training (80%) and testing (20%) sets.
3. Initialize a **Logistic Regression** model using scikit-learn.

In [6]:
from sklearn.model_selection import train_test_split

# 1. Convert categorical features to numbers
# We select our features, including 'Embarked' this time, and convert them to dummy variables
X_categorical = df[['Pclass', 'Sex', 'Age', 'Fare', 'Embarked']]
X_encoded = pd.get_dummies(X_categorical, columns=['Embarked'], drop_first=True)

print("Features after One-Hot Encoding:")
print(X_encoded.columns.tolist())



Features after One-Hot Encoding:
['Pclass', 'Sex', 'Age', 'Fare', 'Embarked_Q', 'Embarked_S']


In [7]:
# 2. Split the data (80% training, 20% testing)
X_train, X_test, y_train, y_test = train_test_split(
    X_encoded, y, test_size=0.2, random_state=42
)

# Verify the split sizes
print(f"\nTraining set size: {X_train.shape[0]} rows")
print(f"Testing set size: {X_test.shape[0]} rows")



Training set size: 712 rows
Testing set size: 179 rows


In [8]:
from sklearn.linear_model import LogisticRegression

# 3. Import and create the model
model = LogisticRegression(max_iter=1000) # max_iter ensures the model has enough attempts to find the best fit
print("\nLogistic Regression Model successfully initialized!")


Logistic Regression Model successfully initialized!


## Step 3: Model Training, Predictions, and Evaluation
In this section, we:
1. Fit (train) our Logistic Regression model using the training data.
2. Generate predictions on the unseen test dataset.
3. Calculate evaluation metrics: Accuracy, Precision, Recall, and F1-Score to analyze performance.
4. Train an alternative **Decision Tree Classifier** to compare the performance of both algorithms.

In [9]:
print("--- Training Logistic Regression ---")
model_lr = LogisticRegression(max_iter=1000)
model_lr.fit(X_train, y_train)

# Predict on test set
lr_preds = model_lr.predict(X_test)

--- Training Logistic Regression ---


In [10]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Calculate metrics
lr_accuracy = accuracy_score(y_test, lr_preds)
lr_precision = precision_score(y_test, lr_preds)
lr_recall = recall_score(y_test, lr_preds)
lr_f1 = f1_score(y_test, lr_preds)

print(f"Logistic Regression Accuracy:  {lr_accuracy:.4f}")
print(f"Logistic Regression Precision: {lr_precision:.4f}")
print(f"Logistic Regression Recall:    {lr_recall:.4f}")
print(f"Logistic Regression F1-Score:  {lr_f1:.4f}")
print("\n" + "="*50 + "\n")

Logistic Regression Accuracy:  0.7989
Logistic Regression Precision: 0.7639
Logistic Regression Recall:    0.7432
Logistic Regression F1-Score:  0.7534




In [11]:
from sklearn.tree import DecisionTreeClassifier

# 2. TRAIN AND EVALUATE DECISION TREE (For Comparison)
print("--- Training Decision Tree Classifier ---")
model_dt = DecisionTreeClassifier(random_state=42, max_depth=5) # max_depth prevents severe overfitting
model_dt.fit(X_train, y_train)

# Predict on test set
dt_preds = model_dt.predict(X_test)

# Calculate metrics
dt_accuracy = accuracy_score(y_test, dt_preds)
dt_precision = precision_score(y_test, dt_preds)
dt_recall = recall_score(y_test, dt_preds)
dt_f1 = f1_score(y_test, dt_preds)

print(f"Decision Tree Accuracy:  {dt_accuracy:.4f}")
print(f"Decision Tree Precision: {dt_precision:.4f}")
print(f"Decision Tree Recall:    {dt_recall:.4f}")
print(f"Decision Tree F1-Score:  {dt_f1:.4f}")

--- Training Decision Tree Classifier ---
Decision Tree Accuracy:  0.7989
Decision Tree Precision: 0.8276
Decision Tree Recall:    0.6486
Decision Tree F1-Score:  0.7273


## Step 4: Model Comparison and Takeaways

* **The Results:** Interestingly, both Logistic Regression and the Decision Tree achieved the exact same overall accuracy (**79.89%**).
* **Under the Hood:** * The **Decision Tree** is highly cautious—it has better **Precision (82.76%)**, meaning its survival predictions are very reliable, but it misses many actual survivors (**Recall: 64.86%**).
  * **Logistic Regression** is the more balanced model, achieving a much better **F1-Score (75.34%)** because its Recall is significantly higher (**74.32%**).
* **Limitations & Next Steps:** Our models are likely held back because continuous features (like `Age` and `Fare`) aren't scaled, and we aren't using engineered features like "Family Size". To improve, I would apply `StandardScaler` to normalize numerical columns and create new features from the raw data.